cpp/src/arrow/util/from_json_test.cc

```diff
@@ -916,7 +914,8 @@ TEST(TestMap, IntegerMapToStringList) {
   auto& key_key_builder = checked_cast<Int16Builder&>(*key_builder.key_builder());
   auto& key_item_builder = checked_cast<Int16Builder&>(*key_builder.item_builder());
   auto& item_builder = checked_cast<ListBuilder&>(*map_builder.item_builder());
-  auto& item_value_builder = checked_cast<StringBuilder&>(*item_builder.value_builder());
+  auto& item_value_builder =
+      checked_cast<class StringBuilder&>(*item_builder.value_builder());
```

Comment:
```suggestion
      checked_cast<StringBuilder&>(*item_builder.value_builder());
```
Where did this change come from?

Reply:
CI failed on this but I lost the error. I'll revert this change so we can see the error in a bit here.

Reply:
I got this to fail locally, I wonder if this is a thing with newer clang?

```
/Library/Developer/CommandLineTools/usr/bin/c++ -DARROW_EXTRA_ERROR_CONTEXT -DARROW_HAVE_NEON -DARROW_WITH_TIMING_TESTS -I/Users/bryce/src/apache/arrow/cpp/build/src -I/Users/bryce/src/apache/arrow/cpp/src -I/Users/bryce/src/apache/arrow/cpp/src/generated -isystem /Users/bryce/src/apache/arrow/cpp/build/_deps/googletest-src/googlemock/include -isystem /Users/bryce/src/apache/arrow/cpp/build/_deps/googletest-src/googletest/include -isystem /Users/bryce/src/apache/arrow/cpp/thirdparty/flatbuffers/include -isystem /Users/bryce/src/apache/arrow/cpp/build/_deps/googletest-src/googletest -isystem /Users/bryce/src/apache/arrow/cpp/build/_deps/googletest-src/googlemock -fno-aligned-new  -Qunused-arguments -fcolor-diagnostics  -Wall -Wextra -Wdocumentation -DARROW_WARN_DOCUMENTATION -Wshorten-64-to-32 -Wno-missing-braces -Wno-unused-parameter -Wno-constant-logical-operand -Wno-return-stack-address -Wdate-time -Wno-unknown-warning-option -Wno-pass-failed -march=armv8-a  -g -Werror -O0 -ggdb  -std=c++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX15.4.sdk -fPIE -MD -MT src/arrow/util/CMakeFiles/arrow-from-json-test.dir/from_json_test.cc.o -MF src/arrow/util/CMakeFiles/arrow-from-json-test.dir/from_json_test.cc.o.d -o src/arrow/util/CMakeFiles/arrow-from-json-test.dir/from_json_test.cc.o -c /Users/bryce/src/apache/arrow/cpp/src/arrow/util/from_json_test.cc
/Users/bryce/src/apache/arrow/cpp/src/arrow/util/from_json_test.cc:918:20: error: must use 'class' tag to refer to type 'StringBuilder' in this scope
  918 |       checked_cast<StringBuilder&>(*item_builder.value_builder());
      |                    ^
      |                    class
```

Reply:
clangd helps explain this here:

```
Must use 'class' tag to refer to type 'StringBuilder' in this scope (fix available)clang(use_of_tag_name_without_tag)
string_builder.h(65, 13): Class 'StringBuilder' is hidden by a non-type declaration of 'StringBuilder' here
```

I think by changing the namespace from arrow::ipc::internal to arrow::util causes a collision with https://github.com/apache/arrow/blob/75acf37683cc32556fbea00771e66eae8af73a42/cpp/src/arrow/util/string_builder.h#L65 so adding `class` helps disambiguate. Is there a better way maybe?

Reply:
For now I put this back the way I had it in b4a9ef4cc75ff6eca671dca60f456b52e05da0b8.

Reply:
Could you use the explicit namespace? (the corresponding function is `arrow::util::StringBuilder` and we should probably rename it too)

```c++
checked_cast<arrow::StringBuilder&>(*item_builder.value_builder());
```

Reply:
We should also make `arrow::util::StringBuilder` internal (in `arrow::internal`) in addition to renaming it, IMHO.

Reply:
> Could you use the explicit namespace? 

Changed to use the fully qualified name in 11da4d40f05e97ad744df58c3d4b0db267e6cbb0. Would it be fair to say that the preferred options in cases like this, in order, are something like,

1. Avoid colliding names in the first place to avoid the need to disambiguate names
2. Use namespaces and fully-qualified names to disambiguate names
3. As a last resort, use keywords like `class` was used here to help the compiler figure it out

Reply:
@bkietz @pitrou I filed https://github.com/apache/arrow/issues/46207 as a follow-on to this PR. In addition to what I describe there, would we also want to rename the files (`string_builder.{h,c}`) when we do that?

Reply:
We can rename the files to mirror the new API names, yes.

Reply:
Sounds good. I'll update the issue body.